utilities/blob_db/blob_db_impl.cc

```diff
@@ -0,0 +1,2202 @@
+  char* buffer = &(*valueptr)[0];
+
+  Slice blob_value;
+  s = reader->Read(handle.offset(), handle.size(), &blob_value, buffer);
+  if (!s.ok() || blob_value.size() != handle.size()) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Failed to read blob from file: %s blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " read: %d key: %s status: '%s'",
+          bfile->PathName().c_str(), handle.offset(), handle.size(),
+          static_cast<int>(blob_value.size()), key.data(),
+          s.ToString().c_str());
+    }
+    return Status::NotFound("Blob Not Found as couldnt retrieve Blob");
+  }
+
+  Slice crc_slice;
+  uint32_t crc_exp;
+  std::string crc_str;
+  crc_str.resize(sizeof(uint32_t));
+  char* crc_buffer = &(crc_str[0]);
+  s = reader->Read(handle.offset() - (key.size() + sizeof(uint32_t)),
+                   sizeof(uint32_t), &crc_slice, crc_buffer);
+  if (!s.ok() || !GetFixed32(&crc_slice, &crc_exp)) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Failed to fetch blob crc file: %s blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s status: '%s'",
+          bfile->PathName().c_str(), handle.offset(), handle.size(), key.data(),
+          s.ToString().c_str());
+    }
+    return Status::NotFound("Blob Not Found as couldnt retrieve CRC");
+  }
+
+  uint32_t crc = crc32c::Extend(0, blob_value.data(), blob_value.size());
+  crc = crc32c::Mask(crc);  // Adjust for storage
+  if (crc != crc_exp) {
+    if (debug_level_ >= 2) {
+      Log(InfoLogLevel::ERROR_LEVEL, db_options_.info_log,
+          "Blob crc mismatch file: %s blob_offset: %" PRIu64
+          " blob_size: %" PRIu64 " key: %s status: '%s'",
+          bfile->PathName().c_str(), handle.offset(), handle.size(), key.data(),
+          s.ToString().c_str());
+    }
+    return Status::Corruption("Corruption. Blob CRC mismatch");
+  }
+
+  if (bdb_options_.compression != kNoCompression) {
+    BlockContents contents;
+    s = UncompressBlockContentsForCompressionType(
+        blob_value.data(), blob_value.size(), &contents,
+        kBlockBasedTableVersionFormat, Slice(), bdb_options_.compression,
+        *(cfd->ioptions()));
+    *value = contents.data.ToString();
+  }
+
+  return s;
+}
+
+Status BlobDBImpl::Get(const ReadOptions& options,
+                       ColumnFamilyHandle* column_family, const Slice& key,
+                       std::string* value) {
+  auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(column_family);
+  auto cfd = cfh->cfd();
+
+  Status s;
+  std::string index_entry;
+  s = db_->Get(options, column_family, key, &index_entry);
+  if (!s.ok()) {
+    if (debug_level_ >= 3)
+      Log(InfoLogLevel::WARN_LEVEL, db_options_.info_log,
+          "Get Failed on LSM KEY: %s status: '%s'", key.ToString().c_str(),
+          s.ToString().c_str());
+    return s;
+  }
+
+  s = CommonGet(cfd, key, index_entry, value);
```

Comment:
I think I asked the question twice.  How do you guarantee the key is not garbage collected between line 1315 and line 1324? Garbage collected because of TTL is fine, but if it is collected because it is overwritten, will we end up with NotFound?